src/main/kotlin/com/defenseunicorns/koci/Repository.kt

```diff
@@ -141,22 +141,40 @@ class Repository(
         }.status.isSuccess()
     }
 
+    suspend fun <T> fetch(descriptor: Descriptor, handler: (stream: InputStream) -> T): T {
+        return client.prepareGet(
+            when (descriptor.mediaType) {
+                MANIFEST_MEDIA_TYPE, INDEX_MEDIA_TYPE -> router.manifest(name, descriptor)
+                else -> router.blob(name, descriptor)
+            }
+        ) {
+            attributes.appendScopes(scopeRepository(name, ACTION_PULL))
+
+            when (descriptor.mediaType) {
+                MANIFEST_MEDIA_TYPE -> accept(ContentType.parse(MANIFEST_MEDIA_TYPE))
+                INDEX_MEDIA_TYPE -> accept(ContentType.parse(INDEX_MEDIA_TYPE))
+            }
+        }.execute { res ->
+            res.body<InputStream>().use { stream ->
+                handler(stream)
+            }
+        }
+    }
+
+    @OptIn(ExperimentalSerializationApi::class)
     suspend fun manifest(descriptor: Descriptor): Result<Manifest> = runCatching {
         require(descriptor.mediaType == MANIFEST_MEDIA_TYPE)
-        val res = client.get(router.manifest(name, descriptor)) {
-            accept(ContentType.parse(MANIFEST_MEDIA_TYPE))
-            attributes.appendScopes(scopeRepository(name, ACTION_PULL))
+        fetch(descriptor) { stream ->
```

Comment:
Bubble up an error if the fetch fails?

Reply:
The `runCatching` will catch the error and bubble into the `Result<T>`

Reply:
Yeah but it will be a pretty generic error right? What's the idiomatic way to report errors for "runCatching"? Just leave it alone or do a custom error?